types/bluebird/index.d.ts

```diff
@@ -490,11 +490,11 @@ declare class Bluebird<R> implements Bluebird.Thenable<R>, Bluebird.Inspection<R
    *
    * *The original object is not modified.*
    */
-  // TODO verify this is correct
+  // TODO disallow non-objects
   // trusted promise for object
-  static props(object: Bluebird<Object>): Bluebird<Object>;
+  static props<T>(object: Bluebird.Thenable<{ [K in keyof T]: Bluebird.Thenable<T[K]> | T[K] }>): Bluebird<{ [K in keyof T]: T[K] }>;
   // object
-  static props(object: Object): Bluebird<Object>;
+  static props<T>(object: { [K in keyof T]: Bluebird.Thenable<T[K]> | T[K] }): Bluebird<{ [K in keyof T]: T[K] }>;
```

Comment:
@DanielRosenwasser was right. I don't really understand how this unrolling works but it's really magical.